### DISCUSSION TITLE
Do I have to create a separate b-tree index for a column I want to search by exact match also?

### DISCUSSION THREAD

**Discussion by GabenGar** [pgroonga/pgroonga#251] [Q&A] [answered]
Let's say I have a table like this:
```pgsql
CREATE TABLE files (
  id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  path text NOT NULL,
  mime_type text NOT NULL
)
```
So I want to search by partial matches but also exact ones for the purpose of grouping by categories (such as `ORDER BY mime_type ASC, id ASC`). Do I need an additional b-tree index for that or just use a different search operator?

**Comment by kou** [comment]
Could you show a `SELECT` what you will try too?

In general, you need multiple indexes for partial match and exact match.

**Reply by GabenGar** [reply]
I wanted to do something like these at first:
```pgsql
-- select all audio files
SELECT
  id
  path
  mime_type
FROM
  files
WHERE
  starts_with(mime_type, 'audio/')
ORDER BY
  id
;
```

```pgsql
-- search for audio files
SELECT
  id
  path
  mime_type
FROM
  files
WHERE
  mime_type &@~ 'audio/'
ORDER BY
  id
;
```
But then I realized it's a bad example, because the table can be rewritten as:
```pgsql
CREATE TABLE files (
  id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  path text NOT NULL,
  type text NOT NULL,
  subtype text NOT NULL
)
```
And then `type`  and `subtype` only need b-tree indexes, while the allowed values are controlled by the application.
I do have an idea for a better example, do I close this discussion and create new one or I just edit the current one?

**Reply by kou** [reply]
Could you close this and create a new one?

**Reply by GabenGar** [reply]
By "close" did you mean accept an answer or outright delete?
The latter can only be done by the admin:
https://docs.github.com/en/discussions/managing-discussions-for-your-community/managing-discussions#deleting-a-discussion

**Reply by kou** [reply]
You can just mark this thread as "answer".

**Reply by GabenGar** [reply]
I'll write a separate reply as an answer.

**Comment by GabenGar** [comment, marked as answer]
In this particular case no pgroonga indexes needed, the table can be rewritten as:
```pgsql
CREATE TABLE files (
  id bigint GENERATED ALWAYS AS IDENTITY PRIMARY KEY,
  path text NOT NULL,
  type text NOT NULL,
  subtype text NOT NULL
);

CREATE INDEX ON files (type);
CREATE INDEX ON files (subtype);
```
The allowed types are known before reaching the database due to [being only a handful of them in the registry](https://www.iana.org/assignments/media-types/media-types.xhtml). The available subtypes are governed by the application config, thus there is no need for partial searches for them in database.
